# vowl Usage Patterns Demo

This notebook demonstrates the different ways to use the **vowl** library for data quality validation against [ODCS](https://github.com/bitol-io/open-data-contract-standard) data contracts.

Each example is split into two cells: an **execution cell** (imports, data wrangling, and the `validate_data(...)` call — copy-paste ready) followed by a **display cell** (the report or DataFrame view). Run the notebook top-to-bottom.

## Contents

**Part 1 · Getting Started**

1. [Setup](#1-setup)

**Part 2 · Running a Validation**

2. [Auto-Mapped Validation](#2-auto-mapped-validation) — Pandas, Polars, other inputs
3. [Explicitly Defined Adapter](#3-explicitly-defined-adapter)
4. [Filtering Rows Before Validation](#4-filtering-rows)

**Part 3 · Understanding the Results**

5. [The ValidationResult Object](#5-validationresult)
6. [Annotated Output — Your Full Table, Flagged](#6-annotated-output)

**Part 4 · Validating Across Multiple Sources**

7. [Multi-Source Validation](#7-multi-source)

**Part 5 · Validating Against Real Databases**

8. [Server-Side Database Validation (Testcontainers)](#9-server-side) — PostgreSQL, MySQL, PySpark, DuckDB ATTACH

---

# Part 1 · Getting Started

<a id="1-setup"></a>
## 1. Setup

Install the dependencies (skipped during execution), then resolve the dataset paths used throughout the notebook.

In [1]:
# (package install cell -- skipped during execution)

In [2]:
from pathlib import Path

# Resolve paths relative to the repo root
REPO_ROOT = Path.cwd()
if REPO_ROOT.name == "examples":
    REPO_ROOT = REPO_ROOT.parent  # examples -> repo root

# Single-source: HDB Resale dataset
HDB_DIR = REPO_ROOT / "tests" / "hdb_resale"
HDB_CSV = HDB_DIR / "HDBResaleWithErrors.csv"
HDB_CONTRACT = HDB_DIR / "hdb_resale_simple.yaml" # Switch out for "hdb_resale.yaml" to test a more complex contract

# Multi-source: Employee dataset (payroll + employee list)
EMPLOYEE_DIR = REPO_ROOT / "tests" / "employee"
EMPLOYEE_PAYROLL_CSV = EMPLOYEE_DIR / "demo_employee_payroll.csv"
EMPLOYEE_LIST_CSV = EMPLOYEE_DIR / "demo_employee_list.csv"
EMPLOYEE_CONTRACT = EMPLOYEE_DIR / "employee_payroll_datacontract.yaml"

# --- Quieten known, benign warnings emitted by this demo ---
import warnings

# vowl surfaces UserWarnings (Arrow type coercion, multi-schema adapter reuse)
# that are informational for this demo dataset.
warnings.filterwarnings("ignore", category=UserWarning,
                        module=r"vowl\.validation\.runner")

In [3]:
# Confirm every dataset path resolves
print(f"HDB CSV exists:          {HDB_CSV.exists()}")
print(f"HDB contract exists:     {HDB_CONTRACT.exists()}")
print(f"Employee payroll CSV exists:  {EMPLOYEE_PAYROLL_CSV.exists()}")
print(f"Employee list CSV exists: {EMPLOYEE_LIST_CSV.exists()}")
print(f"Employee contract exists:     {EMPLOYEE_CONTRACT.exists()}")

HDB CSV exists:          True
HDB contract exists:     True
Employee payroll CSV exists:  True
Employee list CSV exists: True
Employee contract exists:     True


---

# Part 2 · Running a Validation

The ways to feed your data into vowl: let it auto-detect the input type, hand it an explicit adapter, or filter the rows before validating.

<a id="2-auto-mapped-validation"></a>
## 2. Auto-Mapped Validation

When you pass a DataFrame directly to `validate_data(df=...)`, vowl uses a `DataSourceMapper` to automatically detect the input type (pandas, Polars, PyArrow, cuDF, Modin, etc.) and route it to the appropriate adapter -- in most cases, an in-memory DuckDB connection via `IbisAdapter`.

This is the simplest way to validate data. For explicit control over the adapter, see [Section 3](#3-explicitly-defined-adapter).

### 2.1 Pandas

In [4]:
import pandas as pd
from vowl import validate_data

df = pd.read_csv(HDB_CSV, low_memory=False).fillna("")

In [5]:
# Inspect the loaded data
print(f"Loaded {len(df)} rows, {len(df.columns)} columns")
df.head()

Loaded 201879 rows, 11 columns


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price
0,2017-01,ANG MO KIO,2 ROOM,406,ANG MO KIO AVE 10,10 TO 12,44.0,Improved,1979,61 years 04 months,232000.0
1,2017-01,ANG MO KIO,3 ROOM,108,ANG MO KIO AVE 4,01 TO 03,67.0,New Generation,1978,60 years 07 months,250000.0
2,2017-01,ANG MO KIO,3 ROOM,602,ANG MO KIO AVE 5,01 TO 03,67.0,New Generation,1980,62 years 05 months,262000.0
3,2017-01,ANG MO KIO,3 ROOM,465,ANG MO KIO AVE 10,04 TO 06,68.0,New Generation,1980,62 years 01 month,265000.0
4,2017-01,ANG MO KIO,3 ROOM,601,ANG MO KIO AVE 5,01 TO 03,67.0,New Generation,1980,62 years 05 months,265000.0


In [6]:
result = validate_data(contract=str(HDB_CONTRACT), df=df)

In [7]:
# Show the full validation report
result.display_full_report()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       16 / 20 (80.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     201,861 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actua

ValidationResult(passed=False, checks=20, passed_checks=16, failed_checks=4)

<a id="22-polars"></a>
### 2.2 Polars

vowl works with any [Narwhals-compatible](https://github.com/narwhals-dev/narwhals) DataFrame. Here we use Polars with the same contract.

In [8]:
import polars as pl

polars_df = pl.read_csv(HDB_CSV, infer_schema_length=9999999) # To allow mixed type data to be detected in this demo example
result = validate_data(contract=str(HDB_CONTRACT), df=polars_df)

In [9]:
print(f"Loaded {len(polars_df)} rows via Polars")
result.display_full_report()

Loaded 201879 rows via Polars


=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       17 / 20 (85.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       17 / 20 (85.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       17 / 20 (85.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     201,863 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operato

ValidationResult(passed=False, checks=20, passed_checks=17, failed_checks=3)

#### Why Do Pandas and Polars Show Different Failure Counts?

Pandas reports **4 failed checks** (16/20 pass) while Polars reports **3 failed checks** (17/20 pass). The differences stem from how each library handles null/empty values before they reach DuckDB:

| Check | Pandas | Polars | Reason |
|-------|--------|--------|--------|
| **AddressBlockHouseNumber** | FAILED (1 row) | PASSED | Pandas `fillna("")` converts a null `block` to an empty string `""`, which fails the regex `^[A-Za-z0-9]{1,10}$`. Polars keeps it as `null`, and SQL `WHERE ... !~ pattern` evaluates to `NULL` (not `TRUE`), so the row is not counted. |
| **Year** | FAILED (3 rows) | FAILED (2 rows) | One record has an empty `lease_commence_date`. In Pandas, `fillna("")` turns it into `""` which fails the `^[0-9]{4}$` regex. In Polars, the value stays `null` and is skipped by the WHERE clause. |

This is a documented aspect of null handling across backends -- see [Known Issues: Null Handling](../docs/known-issues.md#null-handling-varies-across-database-backends). If you need to catch nulls explicitly, add a `nullValues` library check to your contract.

### 2.3 Other Inputs (Connection Strings, Ibis Backends, Spark, Arrow)

The auto-mapper accepts more than DataFrames. vowl detects the type and creates the appropriate `IbisAdapter` automatically.

In [10]:
# The auto-mapper supports more than just DataFrames.
# Below are other input types that validate_data() can auto-detect:

# --- Connection string (any URI that Ibis recognises) ---
# result = validate_data(contract=str(HDB_CONTRACT), df="postgresql://<user>:<password>@<host>:<port>/<dbname>")
# result = validate_data(contract=str(HDB_CONTRACT), df="snowflake://<user>:<password>@<account>/<db>/<schema>")

# --- Ibis backend (already connected) ---
# con = ibis.duckdb.connect("my_database.ddb")
# result = validate_data(contract=str(HDB_CONTRACT), df=con)

# --- PySpark DataFrame ---
# spark_df = spark.read.table("my_table")
# result = validate_data(contract=str(HDB_CONTRACT), df=spark_df)

# --- PyArrow Table ---
# import pyarrow.csv as pcsv
# arrow_table = pcsv.read_csv("data.csv")
# result = validate_data(contract=str(HDB_CONTRACT), df=arrow_table)

<a id="3-explicitly-defined-adapter"></a>
## 3. Explicitly Defined Adapter

Instead of relying on auto-mapping (Section 2), you can explicitly create an `IbisAdapter` and pass it to `validate_data(adapter=...)`. This gives you full control over the connection, backend, and configuration.

Here we use DuckDB in-memory by materialising the table before validating. The SQL checks run **client-side** within the DuckDB process. This same pattern works with any of the [20+ Ibis backends](https://github.com/ibis-project/ibis) (PostgreSQL, Snowflake, BigQuery, etc.).

In [11]:
import ibis
from vowl import validate_data
from vowl.adapters import IbisAdapter

# Create an in-memory DuckDB connection and register the data
con = ibis.duckdb.connect()
hdb_df = pd.read_csv(HDB_CSV, low_memory=False).fillna("")
hdb_df["lease_commence_date"] = hdb_df["lease_commence_date"].astype(str)
con.create_table("hdb_resale_prices", hdb_df)

# Validate using the IbisAdapter. SQL checks run client-side in DuckDB
adapter = IbisAdapter(con)
result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

In [12]:
# Show the full validation report
result.display_full_report()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       16 / 20 (80.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     201,861 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actua

ValidationResult(passed=False, checks=20, passed_checks=16, failed_checks=4)

<a id="4-filtering-rows"></a>
## 4. Filtering Rows Before Validation

Filter conditions let you validate only a subset of your data, useful for incremental validation of new records.

Table name keys in `filter_conditions` support **wildcard patterns** via Python's [`fnmatch`](https://docs.python.org/3/library/fnmatch.html) module:

| Pattern | Matches | Does Not Match |
|---------|---------|----------------|
| `"emp*"` | `employees`, `emp_history`, `emp_details` | `department` |
| `"table?"` | `table1`, `table2` | `table10` |
| `"[abc]*"` | `accounts`, `billing`, `customers` | `delivery` |
| `"*_archive"` | `orders_archive`, `customers_archive` | `orders` |
| `"*"` | All tables (useful for tenant filtering) | -- |

If multiple patterns match a table, all conditions are combined with AND.

The three examples below share one DuckDB connection.

In [13]:
from vowl.adapters import IbisAdapter, FilterCondition

con = ibis.duckdb.connect()
hdb_df = pd.read_csv(HDB_CSV, low_memory=False).fillna("")
hdb_df["lease_commence_date"] = hdb_df["lease_commence_date"].astype(str)
con.create_table("hdb_resale_prices", hdb_df)

# Dict-style filter: only validate rows from 2024 onwards
adapter = IbisAdapter(
    con,
    filter_conditions={
        "hdb_resale_prices": {
            "field": "month",
            "operator": ">=",
            "value": "2024-01",
        }
    },
)

result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

In [14]:
# Dict-style filter (rows from 2024 onwards)
result.print_summary()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       19 / 20 (95.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       19 / 20 (95.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       19 / 20 (95.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     32,727 / 32,729 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actual 

ValidationResult(passed=False, checks=20, passed_checks=19, failed_checks=1)

In [15]:
# Multiple filter conditions on the same table (combined with AND)
adapter = IbisAdapter(
    con,
    filter_conditions={
        "hdb_resale_prices": [
            FilterCondition(field="month", operator=">=", value="2024-01"),
            FilterCondition(field="town", operator="=", value="ANG MO KIO"),
        ]
    },
)

result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

In [16]:
# Multiple conditions combined with AND
result.print_summary()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       20 / 20 (100.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       20 / 20 (100.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       20 / 20 (100.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     1,264 / 1,264 (100.0%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actua

ValidationResult(passed=True, checks=20, passed_checks=20, failed_checks=0)

In [17]:
# Wildcard pattern matching on table names
# The "*" pattern applies a filter to ALL tables in the contract
adapter = IbisAdapter(
    con,
    filter_conditions={
        # "*" matches every table name (via fnmatch). Useful for multi-tenant
        # scenarios where every table has a shared column like tenant_id.
        "*": FilterCondition(field="month", operator=">=", value="2020-01"),
    },
)

result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

In [18]:
# Wildcard "*" filter applied to every table
result.print_summary()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       19 / 20 (95.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       19 / 20 (95.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       19 / 20 (95.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     137,616 / 137,623 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actua

ValidationResult(passed=False, checks=20, passed_checks=19, failed_checks=1)

---

# Part 3 · Understanding the Results

Every `validate_data(...)` call returns a `ValidationResult`. This part covers the accessors it exposes -- summary DataFrames, failed-row views, and the annotated full table.

<a id="5-validationresult"></a>
## 5. The ValidationResult Object

The `ValidationResult` object provides several ways to inspect and export results. First, run a validation to work with.

In [19]:
# Run a validation to work with
result = validate_data(contract=str(HDB_CONTRACT), df=pd.read_csv(HDB_CSV, low_memory=False).fillna(""))

In [20]:
# Check overall pass/fail, then print the summary (without failed rows)
print(f"All checks passed: {result.passed}")
result.print_summary()

All checks passed: False


=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       16 / 20 (80.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     201,861 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator    

ValidationResult(passed=False, checks=20, passed_checks=16, failed_checks=4)

### Check Results DataFrame

`get_check_results_df()` returns a single DataFrame with one row per check: status, expected/actual values, execution time, dimension, and more.

In [21]:
# Full check results as a DataFrame (one row per check)
check_results_df = result.get_check_results_df()

# To include contract and check definitions in the DataFrame, use:
# check_results_df = result.get_check_results_df(include_contract_definition=True, include_check_definition=True)

In [22]:
# View the check results
check_results_df.to_pandas()

,check_name,target,schema_name,engine,status,operator,actual_value,expected_value,failed_rows_count,aggregation_type,message,rendered_implementation,tables_in_query,check_path,check_ref_type,logical_type,is_generated,execution_time_ms
0,month_column_exists_check,hdb_resale_prices.month,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""month"" FROM ""hdb...",['hdb_resale_prices'],$.schema[0].properties[0].name,DeclaredColumnExistsCheckReference,string,True,5.096542
1,month_logical_type_check,hdb_resale_prices.month,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM ""hdb_resale_prices"" WHERE...",['hdb_resale_prices'],$.schema[0].properties[0].logicalType,LogicalTypeCheckReference,string,True,7.519750
2,town_column_exists_check,hdb_resale_prices.town,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""town"" FROM ""hdb_...",['hdb_resale_prices'],$.schema[0].properties[1].name,DeclaredColumnExistsCheckReference,NaN,True,4.769250
3,flat_type_column_exists_check,hdb_resale_prices.flat_type,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""flat_type"" FROM ...",['hdb_resale_prices'],$.schema[0].properties[2].name,DeclaredColumnExistsCheckReference,NaN,True,4.499875
4,block_column_exists_check,hdb_resale_prices.block,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""block"" FROM ""hdb...",['hdb_resale_prices'],$.schema[0].properties[3].name,DeclaredColumnExistsCheckReference,NaN,True,28.192792
5,street_name_column_exists_check,hdb_resale_prices.street_name,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""street_name"" FRO...",['hdb_resale_prices'],$.schema[0].properties[4].name,DeclaredColumnExistsCheckReference,NaN,True,3.986000
6,storey_range_column_exists_check,hdb_resale_prices.storey_range,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""storey_range"" FR...",['hdb_resale_prices'],$.schema[0].properties[5].name,DeclaredColumnExistsCheckReference,NaN,True,4.176791
7,floor_area_sqm_column_exists_check,hdb_resale_prices.floor_area_sqm,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""floor_area_sqm"" ...",['hdb_resale_prices'],$.schema[0].properties[6].name,DeclaredColumnExistsCheckReference,NaN,True,4.081042
8,flat_model_column_exists_check,hdb_resale_prices.flat_model,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""flat_model"" FROM...",['hdb_resale_prices'],$.schema[0].properties[7].name,DeclaredColumnExistsCheckReference,NaN,True,10.127875
9,lease_commence_date_column_exists_check,hdb_resale_prices.lease_commence_date,hdb_resale_prices,sql,PASSED,mustBe,0,0,0,count,,"SELECT COUNT(*) FROM (SELECT ""lease_commence_d...",['hdb_resale_prices'],$.schema[0].properties[8].name,DeclaredColumnExistsCheckReference,NaN,True,6.002125


### Failed Rows (per check)

`get_output_dfs()` returns a dict of `{check_id: DataFrame}`, one entry per failed check, containing the actual rows that failed.

In [23]:
# Failed rows grouped by check. Each key is "schema::check_name"
output_dfs = result.get_output_dfs()

In [24]:
# Display the failed rows for each check
for check_id, failed_df in output_dfs.items():
    if len(failed_df) == 0:
        continue  # Skip checks with no failed rows
    print(f"\n{'='*60}")
    print(f"Check: {check_id}  ({len(failed_df)} failed rows)")
    print(f"{'='*60}")
    display(failed_df.to_pandas().head())


Check: hdb_resale_prices::AddressBlockHouseNumber  (1 failed rows)


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price,check_id,tables_in_query
0,2017-01,BEDOK,4 ROOM,,BEDOK RESERVOIR RD,07 TO 09,84.0,Simplified,1986,68 years 10 months,395000.0,AddressBlockHouseNumber,hdb_resale_prices



Check: hdb_resale_prices::Month  (2 failed rows)


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price,check_id,tables_in_query
0,2017-jan,BEDOK,5 ROOM,21,CHAI CHEE RD,07 TO 09,130.0,Adjoined flat,1972,54 years 06 months,530000.0,Month,hdb_resale_prices
1,2017-jan,BISHAN,3 ROOM,105,BISHAN ST 12,04 TO 06,4.0,Simplified,1985,67 years 11 months,395000.0,Month,hdb_resale_prices



Check: hdb_resale_prices::Year  (3 failed rows)


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price,check_id,tables_in_query
0,2017-01,ANG MO KIO,3 ROOM,219,ANG MO KIO AVE 1,07 TO 09,67.0,New Generation,1977.0,59 years 06 months,297000.0,Year,hdb_resale_prices
1,2017-01,ANG MO KIO,3 ROOM,211,ANG MO KIO AVE 3,01 TO 03,67.0,New Generation,abc,59 years 03 months,325000.0,Year,hdb_resale_prices
2,2017-01,ANG MO KIO,3 ROOM,330,ANG MO KIO AVE 1,07 TO 09,68.0,New Generation,,63 years,338000.0,Year,hdb_resale_prices



Check: hdb_resale_prices::floor_area_must_be_less_than_200  (12 failed rows)


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price,check_id,tables_in_query
0,2017-06,KALLANG/WHAMPOA,3 ROOM,38,JLN BAHAGIA,01 TO 03,215.0,Terrace,1972,54 years 01 month,830000.0,floor_area_must_be_less_than_200,hdb_resale_prices
1,2017-09,CHOA CHU KANG,EXECUTIVE,641,CHOA CHU KANG ST 64,16 TO 18,215.0,Premium Maisonette,1998,79 years 04 months,888000.0,floor_area_must_be_less_than_200,hdb_resale_prices
2,2017-12,KALLANG/WHAMPOA,3 ROOM,65,JLN MA'MOR,01 TO 03,249.0,Terrace,1972,53 years 07 months,1053888.0,floor_area_must_be_less_than_200,hdb_resale_prices
3,2018-01,CHOA CHU KANG,EXECUTIVE,639,CHOA CHU KANG ST 64,10 TO 12,215.0,Premium Maisonette,1998,79 years,900000.0,floor_area_must_be_less_than_200,hdb_resale_prices
4,2018-09,KALLANG/WHAMPOA,3 ROOM,41,JLN BAHAGIA,01 TO 03,237.0,Terrace,1972,52 years 10 months,1185000.0,floor_area_must_be_less_than_200,hdb_resale_prices


### Consolidated Failed Rows (per table)

`get_consolidated_output_dfs()` deduplicates failed rows across all checks and groups them by table. Useful when multiple checks flag the same row.

In [25]:
# Consolidated: deduplicated failed rows grouped by table
consolidated = result.get_consolidated_output_dfs()

In [26]:
# Display the deduplicated failed rows per table
for table_name, consolidated_df in consolidated.items():
    print(f"\n{'='*60}")
    print(f"Table: {table_name}  ({len(consolidated_df)} unique failed rows)")
    print(f"{'='*60}")
    display(consolidated_df.to_pandas().head())


Table: hdb_resale_prices  (18 unique failed rows)


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price,check_ids,tables_in_query
0,2017-01,BEDOK,4 ROOM,,BEDOK RESERVOIR RD,07 TO 09,84.0,Simplified,1986,68 years 10 months,395000.0,AddressBlockHouseNumber,hdb_resale_prices
1,2017-jan,BEDOK,5 ROOM,21,CHAI CHEE RD,07 TO 09,130.0,Adjoined flat,1972,54 years 06 months,530000.0,Month,hdb_resale_prices
2,2017-jan,BISHAN,3 ROOM,105,BISHAN ST 12,04 TO 06,4.0,Simplified,1985,67 years 11 months,395000.0,Month,hdb_resale_prices
3,2017-01,ANG MO KIO,3 ROOM,219,ANG MO KIO AVE 1,07 TO 09,67.0,New Generation,1977.0,59 years 06 months,297000.0,Year,hdb_resale_prices
4,2017-01,ANG MO KIO,3 ROOM,211,ANG MO KIO AVE 3,01 TO 03,67.0,New Generation,abc,59 years 03 months,325000.0,Year,hdb_resale_prices


### Saving to Disk

`result.save(...)` writes the results (CSVs + a JSON summary) to a directory.

In [27]:
# Save results to disk (CSV + JSON summary)
result.save(output_dir=".", prefix="vowl_demo_HDB_results")


Results saved:
   - vowl_demo_HDB_results_check_results.csv
   - vowl_demo_HDB_results_hdb_resale_prices.csv
   - vowl_demo_HDB_results_summary.json


ValidationResult(passed=False, checks=20, passed_checks=16, failed_checks=4)

<a id="6-annotated-output"></a>
## 6. Annotated Output — Your Full Table, Flagged

The main accessor here is **`get_annotated_output()`**. It returns your full table with one extra column, `check_info`, so you can see failures in the context of every row instead of on their own.

It returns a dict with two keys:

- **`"annotated"`** — `{schema: table}`. Each table is your full data plus a `check_info` column. Every original row is kept. `check_info` is `null` for rows that passed every check, and a list of the failing check(s) for rows that didn't.
- **`"residues"`** — failed rows for checks that can't be attached to a single table (cross-table, aggregation, and column-subset checks). The single-table HDB contract below has none; the [residual-rows example](#residual-rows) at the end of this section shows them using a multi-source contract.

> Not every check can be merged into the annotated table. For the full rules and worked examples, see [Known Issues: Annotated Output](../docs/known-issues.md#annotated-output-not-all-checks-can-be-merged).

### Choosing How Much Detail: the `check_info` Options

`check_info` is a **list of objects, one per failing check** (stored as a JSON string in the column). The `check_info=` argument sets how many fields each object carries:

| `check_info` | Each object contains | Use it when |
|---|---|---|
| `"names"` *(default)* | `check_name` | You only need to know *which* checks a row failed |
| `"summary"` | `check_name`, `dimension`, `tags`, `target` | You want each failure's quality dimension, tags, and column — e.g. to split completeness vs. accuracy |
| `"full"` | the entire check definition (`type`, `description`, `query`, `mustBe`, `dimension`, `tags`) plus `check_name` and `target` | You want everything, including the underlying rule/SQL, on every row |

All three return the **same shape** — a list of objects — so you always read a value the same way (`json.loads(cell)[0]["check_name"]`); they differ only in how many fields each object has. A row failing two checks gets a two-object list.

```text
"names"    [{"check_name": "Month"}]
"summary"  [{"check_name": "Month", "dimension": "conformity",
             "tags": ["SG-DRM v5.0"], "target": "hdb_resale_prices.month"}]
"full"     [{"type": "sql", "name": "Month", "description": "...", "mustBe": 0,
             "query": "SELECT COUNT(*) ...", "tags": ["SG-DRM v5.0"],
             "dimension": "conformity", "check_name": "Month",
             "target": "hdb_resale_prices.month"}]
```

The examples below use `"summary"`.

In [28]:
# "annotated" -> {schema: your full table + a check_info column}
# "residues"  -> failed rows for checks that can't attach to one table (shown later)
# check_info="summary" -> each object also carries dimension, tags, and target.
output = result.get_annotated_output(check_info="summary")
annotated = output["annotated"]["hdb_resale_prices"].to_pandas()

In [29]:
# Inspect the structure of the annotated output
print("Top-level keys:    ", list(output.keys()))
print("Annotated schemas: ", list(output["annotated"].keys()))
print("Residue keys:      ", list(output["residues"].keys()))
print(f"\nAnnotated table: {annotated.shape[0]:,} rows x {annotated.shape[1]} columns")
print("Columns:", list(annotated.columns))

Top-level keys:     ['annotated', 'residues']
Annotated schemas:  ['hdb_resale_prices']
Residue keys:       []

Annotated table: 201,879 rows x 12 columns
Columns: ['month', 'town', 'flat_type', 'block', 'street_name', 'storey_range', 'floor_area_sqm', 'flat_model', 'lease_commence_date', 'remaining_lease', 'resale_price', 'check_info']


### Inspecting the Full Table, Flagged Rows First

Every original row is present. `check_info` is `null` for rows that passed and a JSON list for rows that failed — one object per failing check (here carrying `dimension`, `tags`, and `target` from the `"summary"` preset). Sorting by `check_info` (nulls last) floats the flagged rows to the top while still showing the whole table.

In [30]:
# Sort the full table so flagged rows surface first; passing rows (check_info = null) sink.
flagged_first = annotated.sort_values("check_info", na_position="last").reset_index(drop=True)
n_flagged = annotated["check_info"].notna().sum()

In [31]:
print(f"{n_flagged} flagged rows at the top, {len(annotated) - n_flagged:,} passing rows below\n")
flagged_first[["month", "town", "block", "floor_area_sqm",
               "lease_commence_date", "check_info"]].head(20)

18 flagged rows at the top, 201,861 passing rows below



,month,town,block,floor_area_sqm,lease_commence_date,check_info
0,2017-01,BEDOK,,84.0,1986,"[{""check_name"": ""AddressBlockHouseNumber"", ""di..."
1,2017-jan,BEDOK,21,130.0,1972,"[{""check_name"": ""Month"", ""dimension"": ""conform..."
2,2017-jan,BISHAN,105,4.0,1985,"[{""check_name"": ""Month"", ""dimension"": ""conform..."
3,2017-01,ANG MO KIO,219,67.0,1977.0,"[{""check_name"": ""Year"", ""dimension"": ""conformi..."
4,2017-01,ANG MO KIO,211,67.0,abc,"[{""check_name"": ""Year"", ""dimension"": ""conformi..."
5,2017-01,ANG MO KIO,330,68.0,,"[{""check_name"": ""Year"", ""dimension"": ""conformi..."
6,2017-06,KALLANG/WHAMPOA,38,215.0,1972,"[{""check_name"": ""floor_area_must_be_less_than_..."
7,2017-09,CHOA CHU KANG,641,215.0,1998,"[{""check_name"": ""floor_area_must_be_less_than_..."
8,2017-12,KALLANG/WHAMPOA,65,249.0,1972,"[{""check_name"": ""floor_area_must_be_less_than_..."
9,2018-01,CHOA CHU KANG,639,215.0,1998,"[{""check_name"": ""floor_area_must_be_less_than_..."


### Keeping the Clean Rows for Downstream Use

Because the annotation lives on the full table, separating the good rows from the bad is a one-liner. Filter to where `check_info` is null, drop the annotation column, and you have a clean dataset ready to feed downstream -- no separate join back to the source needed.

In [32]:
# Keep only the rows that passed every check, then drop the annotation column
clean = annotated[annotated["check_info"].isna()].drop(columns=["check_info"])

In [33]:
print(f"Clean rows ready for downstream use: {len(clean):,} of {len(annotated):,}")
clean.head()

Clean rows ready for downstream use: 201,861 of 201,879


,month,town,flat_type,block,street_name,storey_range,floor_area_sqm,flat_model,lease_commence_date,remaining_lease,resale_price
0,2017-01,ANG MO KIO,2 ROOM,406,ANG MO KIO AVE 10,10 TO 12,44.0,Improved,1979,61 years 04 months,232000.0
1,2017-01,ANG MO KIO,3 ROOM,108,ANG MO KIO AVE 4,01 TO 03,67.0,New Generation,1978,60 years 07 months,250000.0
2,2017-01,ANG MO KIO,3 ROOM,602,ANG MO KIO AVE 5,01 TO 03,67.0,New Generation,1980,62 years 05 months,262000.0
3,2017-01,ANG MO KIO,3 ROOM,465,ANG MO KIO AVE 10,04 TO 06,68.0,New Generation,1980,62 years 01 month,265000.0
4,2017-01,ANG MO KIO,3 ROOM,601,ANG MO KIO AVE 5,01 TO 03,67.0,New Generation,1980,62 years 05 months,265000.0


<a id="residual-rows"></a>
### Residual (Non-Mergeable) Rows

The HDB examples above are single-table, so `residues` is empty. Residues appear when a check can't be attached to one table — **cross-table** checks (e.g. a reference between two schemas), **aggregation** checks, and **column-subset** checks.

Residues are **per-check**: `get_annotated_output()` returns one entry for each non-mergeable check, keyed `"<schema>::<check_name>"`. They are never grouped together, so a row that fails two such checks appears once under each check's entry. Each entry carries the failed rows plus a single-name `check_ids` column and `tables_in_query` (not `check_info`).

Below is a quick multi-source run on the Employee dataset, whose contract has cross-table checks. (Multi-source validation is covered in [Section 7](#7-multi-source); we borrow it here just to produce residues.)

In [34]:
import ibis
from vowl.adapters import IbisAdapter

# Load the two Employee tables onto one DuckDB connection (cross-table contract)
con = ibis.duckdb.connect()
con.create_table("demo_employee_payroll", pd.read_csv(EMPLOYEE_PAYROLL_CSV))
con.create_table("demo_employee_list", pd.read_csv(EMPLOYEE_LIST_CSV))

mt_result = validate_data(contract=str(EMPLOYEE_CONTRACT), adapter=IbisAdapter(con))
mt_output = mt_result.get_annotated_output()

In [35]:
# Each residue is the failed rows for ONE non-mergeable check, keyed "<schema>::<check>"
print("Annotated schemas:", list(mt_output["annotated"].keys()))
print("Residue keys:     ", list(mt_output["residues"].keys()))

for key, residue in mt_output["residues"].items():
    residue_df = residue.to_pandas()
    print(f"\nResidue '{key}': {len(residue_df)} failed row(s)")
    display(residue_df[["employee_id", "payroll_id", "month",
                        "check_ids", "tables_in_query"]])

Annotated schemas: ['demo_employee_payroll', 'demo_employee_list']
Residue keys:      ['demo_employee_payroll::employee_id_exists_in_master_list', 'demo_employee_payroll::phone_number_exists_in_master_list']

Residue 'demo_employee_payroll::employee_id_exists_in_master_list': 1 failed row(s)


,employee_id,payroll_id,month,check_ids,tables_in_query
0,e939123,e52e556f-79b0-471f-ad08-e27b2c524ace,2025-12,employee_id_exists_in_master_list,"demo_employee_list, demo_employee_payroll"



Residue 'demo_employee_payroll::phone_number_exists_in_master_list': 2 failed row(s)


,employee_id,payroll_id,month,check_ids,tables_in_query
0,e128903,cb04c5bb-9386-44cf-a565-2276744c9cc0,2025-12,phone_number_exists_in_master_list,"demo_employee_list, demo_employee_payroll"
1,e939123,e52e556f-79b0-471f-ad08-e27b2c524ace,2025-12,phone_number_exists_in_master_list,"demo_employee_list, demo_employee_payroll"


### Writing Annotated Tables to Disk

`result.save(...)` can write the annotated tables instead of (or alongside) the failed-rows CSVs. The `output_mode` argument controls the layout (every mode also writes `<prefix>_check_results.csv` and `<prefix>_summary.json`):

| Mode | Files written |
|------|----------------|
| `"failed_rows"` (default) | One grouped failed-rows CSV per table key, e.g. `<prefix>_<table>.csv` |
| `"annotated"` | One `<prefix>_<schema>_annotated.csv` per schema (full table + `check_info`), **plus** one `<prefix>_<schema>_<check>_residue.csv` per non-mergeable check |
| `"both"` | The failed-rows CSVs **and** the annotated CSVs (residues are already covered by the grouped failed-rows CSVs, so no separate `_residue` files are written) |

The `check_info` argument (`"names"` / `"summary"` / `"full"`, described above) sets how much detail the saved `check_info` column carries. To make annotated output the default for every `save()`, pass `ValidationConfig(output_mode="annotated", annotated_check_info="summary")` to `validate_data(config=...)`.

Single-table contracts (like the HDB example) have no non-mergeable checks, so they never produce `_residue` files. The multi-source run below does — let's save it to see them.

In [36]:
# Write the annotated table(s) to disk
result.save(output_dir=".", prefix="vowl_demo_annotated", output_mode="annotated")


Results saved:
   - vowl_demo_annotated_check_results.csv
   - vowl_demo_annotated_hdb_resale_prices_annotated.csv
   - vowl_demo_annotated_summary.json


ValidationResult(passed=False, checks=20, passed_checks=16, failed_checks=4)

#### Saving a Run That Has Residues

`result` above is single-table, so its `save(output_mode="annotated")` wrote only `*_annotated.csv` files. Saving the multi-source `mt_result` from the residue example instead also writes one `*_residue.csv` per non-mergeable check — keyed `<schema>_<check>` (the `::` in the residue key becomes `_`). Each residue CSV carries that one check's failed rows plus a single-name `check_ids` column and `tables_in_query`.

In [37]:
import os

# Save the multi-source run; annotated mode emits residue CSVs for the
# cross-table checks that can't be merged onto a single table.
mt_result.save(output_dir=".", prefix="vowl_demo_residues", output_mode="annotated")

print("\nFiles written:")
for fname in sorted(f for f in os.listdir(".") if f.startswith("vowl_demo_residues")):
    tag = "  <- residue" if fname.endswith("_residue.csv") else ""
    print(f"   {fname}{tag}")


Results saved:
   - vowl_demo_residues_check_results.csv
   - vowl_demo_residues_demo_employee_payroll_annotated.csv
   - vowl_demo_residues_demo_employee_list_annotated.csv
   - vowl_demo_residues_demo_employee_payroll_employee_id_exists_in_master_list_residue.csv
   - vowl_demo_residues_demo_employee_payroll_phone_number_exists_in_master_list_residue.csv
   - vowl_demo_residues_summary.json

Files written:
   vowl_demo_residues_check_results.csv
   vowl_demo_residues_demo_employee_list_annotated.csv
   vowl_demo_residues_demo_employee_payroll_annotated.csv
   vowl_demo_residues_demo_employee_payroll_employee_id_exists_in_master_list_residue.csv  <- residue
   vowl_demo_residues_demo_employee_payroll_phone_number_exists_in_master_list_residue.csv  <- residue
   vowl_demo_residues_summary.json


### Where the Full Check Definitions Live

To keep the annotated table readable, you'll usually pick `"names"` or `"summary"` rather than `"full"`. You don't lose anything: the complete definition of **every** check — its `dimension`, `tags`, `description`, `query`, and more — is always written to the `*_summary.json` file that `save()` produces, keyed by check name. Keep the table light, and look up the rest here when you need it.

In [38]:
import json

# save(prefix="vowl_demo_annotated", ...) wrote this alongside the CSVs.
with open("vowl_demo_annotated_summary.json") as f:
    summary = json.load(f)

# Every check's resolved definition lives under "check_results", keyed by name.
month_check = next(c for c in summary["check_results"] if c["name"] == "Month")

print("Target:    ", month_check["target"])
print("Dimension: ", month_check["check_definition"]["dimension"])
print("Tags:      ", month_check["check_definition"]["tags"])
print("\nFull check_definition:")
print(json.dumps(month_check["check_definition"], indent=2))

Target:     hdb_resale_prices.month
Dimension:  conformity
Tags:       ['SG-DRM v5.0']

Full check_definition:
{
  "type": "sql",
  "name": "Month",
  "description": "Based on ISO 8601, assumed to be in UTC +8 | YYYY-MM",
  "mustBe": 0,
  "query": "SELECT COUNT(*)\nFROM \"hdb_resale_prices\"\nWHERE CAST(month AS TEXT) !~ '^[0-9]{4}-(0[1-9]|1[0-2])$';",
  "tags": [
    "SG-DRM v5.0"
  ],
  "dimension": "conformity"
}


---

# Part 4 · Validating Across Multiple Sources

Contracts can span several tables -- even tables living in different source systems. This part covers multi-source validation and the residual rows produced by checks that span more than one table.

<a id="7-multi-source"></a>
## 7. Multi-Source Validation

Validate across tables that live in different source systems. The Employee contract has two schemas, `demo_employee_payroll` and `demo_employee_list`, with cross-table referential integrity checks (e.g. "every payroll employee_id must exist in the master list").

> **Note:** Multi-adapter usage loads data into a local DuckDB instance before running quality checks. Ensure the local machine can handle the combined data size.

In [39]:
import ibis
from vowl import validate_data
from vowl.adapters import IbisAdapter

# Load each table into its own connection (simulating different source systems)
payroll_df = pd.read_csv(EMPLOYEE_PAYROLL_CSV)
employee_list_df = pd.read_csv(EMPLOYEE_LIST_CSV)

con_payroll = ibis.duckdb.connect()
con_payroll.create_table("demo_employee_payroll", payroll_df)

con_employees = ibis.duckdb.connect()
con_employees.create_table("demo_employee_list", employee_list_df)

# Map each schema name to its adapter
adapters = {
    "demo_employee_payroll": IbisAdapter(con_payroll),
    "demo_employee_list": IbisAdapter(con_employees),
}

result = validate_data(contract=str(EMPLOYEE_CONTRACT), adapters=adapters)

In [40]:
# Show the full validation report
result.display_full_report()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           b6376e57-aa90-41eb-90d0-c72dca7567e6
   Schemas:               demo_employee_payroll, demo_employee_list

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       70 / 76 (92.1%)

   demo_employee_payroll:
     Overall:
       Checks Pass Rate:       48 / 53 (90.5%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       48 / 51 (94.1%)
       ERRORED Checks:         0
       Unique Passed Rows:     0 / 3 (0.0%)
     Multi Table:
       Checks Pass Rate:       0 / 2 (0.0%)
       ERRORED Checks:         0
       Non-unique Failed Rows: 3

   demo_employee_list:
     Overall:
       Checks Pass Rate:       22 / 23 (95.6%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       22 / 23 (95.6%)
       ERRORED Checks:         0
       Unique Passed Rows:     0 / 2 (0.0%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)


ValidationResult(passed=False, checks=76, passed_checks=70, failed_checks=6)

In [41]:
# Save the multi-table results to disk
result.save(output_dir=".", prefix="vowl_demo_multi_table_results")


Results saved:
   - vowl_demo_multi_table_results_check_results.csv
   - vowl_demo_multi_table_results_demo_employee_list.csv
   - vowl_demo_multi_table_results_demo_employee_list_demo_employee_payroll.csv
   - vowl_demo_multi_table_results_demo_employee_payroll.csv
   - vowl_demo_multi_table_results_summary.json


ValidationResult(passed=False, checks=76, passed_checks=70, failed_checks=6)

### Single Adapter Expanding to All Schemas

When all tables live on the same connection, pass a single `adapter`; vowl will reuse it across all schemas in the contract.

In [42]:
# Both tables on the same DuckDB connection
con = ibis.duckdb.connect()
con.create_table("demo_employee_payroll", payroll_df)
con.create_table("demo_employee_list", employee_list_df)

adapter = IbisAdapter(con)

# vowl warns that a single adapter is being expanded to multiple schemas
result = validate_data(contract=str(EMPLOYEE_CONTRACT), adapter=adapter)

In [43]:
# Show the full validation report
result.display_full_report()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           b6376e57-aa90-41eb-90d0-c72dca7567e6
   Schemas:               demo_employee_payroll, demo_employee_list

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       70 / 76 (92.1%)

   demo_employee_payroll:
     Overall:
       Checks Pass Rate:       48 / 53 (90.5%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       48 / 51 (94.1%)
       ERRORED Checks:         0
       Unique Passed Rows:     0 / 3 (0.0%)
     Multi Table:
       Checks Pass Rate:       0 / 2 (0.0%)
       ERRORED Checks:         0
       Non-unique Failed Rows: 3

   demo_employee_list:
     Overall:
       Checks Pass Rate:       22 / 23 (95.6%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       22 / 23 (95.6%)
       ERRORED Checks:         0
       Unique Passed Rows:     0 / 2 (0.0%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)


ValidationResult(passed=False, checks=76, passed_checks=70, failed_checks=6)

---

# Part 5 · Validating Against Real Databases

The examples below use [testcontainers](https://testcontainers-python.readthedocs.io/) to spin up real database instances in Docker. Unlike Part 2 where checks run client-side in DuckDB, here the SQL checks execute **on the database server itself** (PostgreSQL, MySQL, PySpark, etc.).

**Requirements:**

- Docker must be running
- Dev dependencies: `uv sync --dev`

<a id="9-server-side"></a>
## 9. Server-Side Database Validation (Testcontainers)

### 9.1 PostgreSQL

In [44]:
import os
from pathlib import Path as _Path

import ibis
import pandas as pd
from testcontainers.postgres import PostgresContainer
from vowl import validate_data
from vowl.adapters import IbisAdapter

# --- Docker / Infrastructure Setup ---
# Configure Docker Desktop socket path (macOS)
docker_sock = _Path.home() / ".docker" / "run" / "docker.sock"
if docker_sock.exists() and "DOCKER_HOST" not in os.environ:
    os.environ["DOCKER_HOST"] = f"unix://{docker_sock}"

# Disable Ryuk reaper for Docker Desktop compatibility
os.environ.setdefault("TESTCONTAINERS_RYUK_DISABLED", "true")

postgres = PostgresContainer("postgres:15-alpine")
postgres.start()

# Connect via Ibis
pg_con = ibis.postgres.connect(
    host=postgres.get_container_host_ip(),
    port=postgres.get_exposed_port(5432),
    user=postgres.username,
    password=postgres.password,
    database=postgres.dbname,
)

# Load HDB data as TEXT for PostgreSQL
hdb_df = pd.read_csv(HDB_CSV, low_memory=False).fillna("").astype(str)

pg_con.raw_sql("""
    CREATE TABLE hdb_resale_prices (
        month TEXT, town TEXT, flat_type TEXT, block TEXT,
        street_name TEXT, storey_range TEXT, floor_area_sqm TEXT,
        flat_model TEXT, lease_commence_date TEXT, remaining_lease TEXT,
        resale_price TEXT
    )
""")
pg_con.insert("hdb_resale_prices", hdb_df)
print("PostgreSQL container started and data loaded.")

PostgreSQL container started and data loaded.


In [45]:
# --- Validation (runs server-side on PostgreSQL) ---
adapter = IbisAdapter(pg_con)
result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

In [46]:
# Show the report, then tear down the container
result.display_full_report()
postgres.stop()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       15 / 20 (75.0%)
     ERRORED Checks:         2

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       15 / 20 (75.0%)
       ERRORED Checks:         2
     Single Table:
       Checks Pass Rate:       15 / 20 (75.0%)
       ERRORED Checks:         2
       Unique Passed Rows:     201,873 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operat

> **Why do 2 checks ERROR on PostgreSQL but not on MySQL?**
>
> Both backends store all columns as TEXT, but the checks `floor_area_must_be_less_than_200` and `resale_price_must_not_exceed_2m` compare columns to numeric literals (e.g. `WHERE floor_area_sqm >= 200`). vowl's `apply_try_cast` transform wraps the column in a safe cast, but the rendered SQL differs by backend:
>
> | Backend | Rendered SQL | Behaviour |
> |---------|-------------|-----------|
> | **MySQL** | `CAST(floor_area_sqm AS SIGNED) >= 200` | MySQL truncates `"130.0"` to `130` -- no error |
> | **PostgreSQL** | `CAST(floor_area_sqm AS BIGINT) >= 200` | PostgreSQL has no `TRY_CAST`, so sqlglot downgrades to strict `CAST`, which rejects `"130.0"` as invalid integer syntax |
>
> In a production PostgreSQL setup, you would use proper column types (`INTEGER`, `NUMERIC`), which would:
> 1. **Reject invalid data at INSERT time** -- values like `"abc"` or `""` would never enter the table
> 2. **Make the CAST unnecessary** -- the column is already numeric, so `WHERE floor_area_sqm >= 200` executes directly
>
> See [Known Issues](../docs/known-issues.md) for more on backend-specific dialect differences.

### 9.2 MySQL

> **Note:** For MySQL, select the database when creating the connection (via `database=` param or in the connection URI). vowl does not issue `USE database`; it runs read-only `SELECT` queries against the active database.

In [47]:
from testcontainers.mysql import MySqlContainer

# --- Docker / Infrastructure Setup ---
mysql = MySqlContainer("mysql:8.0")
mysql.start()

mysql_con = ibis.mysql.connect(
    host=mysql.get_container_host_ip(),
    port=int(mysql.get_exposed_port(3306)),
    user=mysql.username,
    password=mysql.password,
    database=mysql.dbname,
)

# Load data. MySQL requires TEXT types for string columns
hdb_df = pd.read_csv(HDB_CSV, low_memory=False).fillna("").astype(str)

mysql_con.raw_sql("""
    CREATE TABLE hdb_resale_prices (
        month TEXT, town TEXT, flat_type TEXT, block TEXT,
        street_name TEXT, storey_range TEXT, floor_area_sqm TEXT,
        flat_model TEXT, lease_commence_date TEXT, remaining_lease TEXT,
        resale_price TEXT
    )
""")
mysql_con.insert("hdb_resale_prices", hdb_df)
print("MySQL container started and data loaded.")

MySQL container started and data loaded.


In [48]:
# --- Validation (runs server-side on MySQL) ---
adapter = IbisAdapter(mysql_con)
result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

In [49]:
# Show the report, then tear down the container
result.display_full_report()
mysql.stop()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       16 / 20 (80.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       16 / 20 (80.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     201,861 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actua

### 9.3 PySpark

PySpark validation uses `ibis.pyspark.connect()` to bridge Spark SQL with vowl's `IbisAdapter`. The SQL checks execute within the Spark engine.

**Requirements:**

- PySpark installed (`pip install pyspark`)
- Java 11+ available (set `JAVA_HOME` if needed)

In [50]:
import os

# Auto-detect Java on macOS (Homebrew) if JAVA_HOME not set
if "JAVA_HOME" not in os.environ:
    from pathlib import Path as _P
    homebrew_java = _P("/opt/homebrew/opt/openjdk@17")
    if homebrew_java.exists():
        os.environ["JAVA_HOME"] = str(homebrew_java)

from pyspark.sql import SparkSession

# --- Spark Setup ---
spark = (
    SparkSession.builder
    .master("local[1]")
    .appName("vowl_demo")
    .config("spark.driver.memory", "512m")
    .config("spark.sql.shuffle.partitions", "1")
    .getOrCreate()
)
spark.sparkContext.setLogLevel("ERROR")

# Load the data as a Spark DataFrame
spark_df = spark.createDataFrame(
    pd.read_csv(HDB_CSV, low_memory=False).fillna("").astype(str)
)
spark_df.createOrReplaceTempView("hdb_resale_prices")

# --- Validation (runs within the Spark engine) ---
con = ibis.pyspark.connect(session=spark)
adapter = IbisAdapter(con)
result = validate_data(contract=str(HDB_CONTRACT), adapter=adapter)

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/06/28 21:07:30 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Exception ignored while finalizing file <_io.BufferedWriter name=5>:
Traceback (most recent call last):
  File "/Users/jamestth/DEP/DCP/dqmk/.venv/lib/python3.14/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


In [51]:
# Show the report, then stop the Spark session
result.display_full_report()
spark.stop()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           c11443ee-542f-4442-b28d-2d224342be37
   Schemas:               hdb_resale_prices

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       17 / 20 (85.0%)

   hdb_resale_prices:
     Overall:
       Checks Pass Rate:       17 / 20 (85.0%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       17 / 20 (85.0%)
       ERRORED Checks:         0
       Unique Passed Rows:     201,873 / 201,879 (99.9%)
     Multi Table:
       Checks Pass Rate:       0 / 0 (N/A)
       ERRORED Checks:         0
       Non-unique Failed Rows: 0


 CHECK RESULTS
+-----------------------------------------+---------------------------------------+-------------------+--------+---------------+---------------+--------+----------------+
| check_id                                | Target                                | tables_in_query   | status | operator      | expected      | actua

### 9.4 DuckDB ATTACH (Multi-Source Workaround)

As shown in 9.1, some backends produce `ERROR` checks due to dialect limitations (e.g. PostgreSQL lacks `TRY_CAST`, MSSQL lacks regex). A workaround is to **attach** the remote database to a local DuckDB instance and run all queries through DuckDB's engine instead.

This gives you DuckDB's full SQL support (`TRY_CAST`, `REGEXP_MATCHES`, etc.) while reading data directly from the remote database -- no manual data loading required. Here we use the Employee dataset to demonstrate cross-table validation through DuckDB ATTACH.

> **Note:** DuckDB ATTACH currently supports PostgreSQL, MySQL, and SQLite. See [Known Issues](../docs/known-issues.md#mssql-no-regex-support) for the MSSQL workaround using the same pattern.

In [52]:
# --- Simulate two separate source systems, each with its own database ---
pg_payroll = PostgresContainer("postgres:15-alpine")
pg_payroll.start()

pg_employees = PostgresContainer("postgres:15-alpine")
pg_employees.start()

# Load payroll data into the first database
payroll_con = ibis.postgres.connect(
    host=pg_payroll.get_container_host_ip(),
    port=pg_payroll.get_exposed_port(5432),
    user=pg_payroll.username, password=pg_payroll.password, database=pg_payroll.dbname,
)
payroll_con.create_table("demo_employee_payroll", pd.read_csv(EMPLOYEE_PAYROLL_CSV).astype(str))

# Load employee list into the second database
employees_con = ibis.postgres.connect(
    host=pg_employees.get_container_host_ip(),
    port=pg_employees.get_exposed_port(5432),
    user=pg_employees.username, password=pg_employees.password, database=pg_employees.dbname,
)
employees_con.create_table("demo_employee_list", pd.read_csv(EMPLOYEE_LIST_CSV).astype(str))
print("Two PostgreSQL containers started (payroll + employee list).")

# --- DuckDB ATTACH: bridge both databases into one DuckDB connection ---
duck_con = ibis.duckdb.connect()
duck_con.raw_sql(f"""
    ATTACH 'dbname={pg_payroll.dbname} host={pg_payroll.get_container_host_ip()}
           port={pg_payroll.get_exposed_port(5432)}
           user={pg_payroll.username} password={pg_payroll.password}'
    AS payroll_db (TYPE postgres, READ_ONLY)
""")
duck_con.raw_sql(f"""
    ATTACH 'dbname={pg_employees.dbname} host={pg_employees.get_container_host_ip()}
           port={pg_employees.get_exposed_port(5432)}
           user={pg_employees.username} password={pg_employees.password}'
    AS employees_db (TYPE postgres, READ_ONLY)
""")

# Attached tables are namespaced (e.g. payroll_db.public.demo_employee_payroll).
# Create views as prefix-free shortcuts so contract queries can reference tables directly.
duck_con.raw_sql("USE memory")
duck_con.raw_sql("CREATE VIEW demo_employee_payroll AS SELECT * FROM payroll_db.public.demo_employee_payroll")
duck_con.raw_sql("CREATE VIEW demo_employee_list AS SELECT * FROM employees_db.public.demo_employee_list")

# Validate -- cross-table checks (e.g. referential integrity between payroll
# and employee list) work because both views are on the same DuckDB connection.
adapter = IbisAdapter(duck_con)
result = validate_data(contract=str(EMPLOYEE_CONTRACT), adapter=adapter)

Two PostgreSQL containers started (payroll + employee list).


In [53]:
# Show the report, then tear down both containers
result.display_full_report()
pg_payroll.stop()
pg_employees.stop()



=== Data Quality Validation Results ===
   Contract Version:      v3.1.0
   Contract ID:           b6376e57-aa90-41eb-90d0-c72dca7567e6
   Schemas:               demo_employee_payroll, demo_employee_list

 OVERALL DATA QUALITY
   Overall:
     Checks Pass Rate:       68 / 76 (89.4%)
     ERRORED Checks:         2

   demo_employee_payroll:
     Overall:
       Checks Pass Rate:       46 / 53 (86.7%)
       ERRORED Checks:         2
     Single Table:
       Checks Pass Rate:       46 / 51 (90.1%)
       ERRORED Checks:         2
       Unique Passed Rows:     0 / 3 (0.0%)
     Multi Table:
       Checks Pass Rate:       0 / 2 (0.0%)
       ERRORED Checks:         0
       Non-unique Failed Rows: 3

   demo_employee_list:
     Overall:
       Checks Pass Rate:       22 / 23 (95.6%)
       ERRORED Checks:         0
     Single Table:
       Checks Pass Rate:       22 / 23 (95.6%)
       ERRORED Checks:         0
       Unique Passed Rows:     0 / 2 (0.0%)
     Multi Table:
       Check